### PR TITLE
Ensure empty constructors return 0-length vectors

### LINKED ISSUE #707: Zero-input constructors should return zero-length objects
Rather than an error, as currently:

``` r
library(lubridate)

duration()
#> Error in duration(): No valid values have been passed to 'duration' constructor
period()
#> Error in period(): No valid values have been passed to 'period' constructor
interval()
#> Error in tz(start): argument "start" is missing, with no default
```

This makes more similar in behaviour to base R

```r
integer()
#> integer(0)
factor()
#> factor(0)
#> Levels:
```

And this property is taken advantage of in [vctrs](http://vctrs.r-lib.org)

### PR DESCRIPTION
**Pull request by hadley**
Fixes #707

### CODE REVIEW

File: R/periods.r [hunk 1]
```diff
@@ -393,7 +400,12 @@ setMethod("$<-", signature(x = "Period"), function(x, name, value) {
 #' @export
 period <- function(num = NULL, units = "second", ...) {
   nums <- list(...)
-  if (is.character(num)) {
+  if (is.null(num) && missing(...)) {
```
Comment (vspinu) [inline R/periods.r]: I think this change and the entire spaghetti of the if-elses that follows can be reduced to one if-else with the pre-conversion `num <- as.numeric(num)`. Let me try it out. 

BTW, `period(numeric())` fails ATM.

**Comment by vspinu** [comment]
I have just realized that we have different contracts for legacy duration and period constructors. One is vectorised another is not:

```R
period(c(30, 20), units = c("sec", "days"))
[1] "20d 0H 0M 30S"
> duration(c(30, 20), units = c("sec", "days"))
[1] "30s"                    "1728000s (~2.86 weeks)"
```

**Comment by vspinu** [comment]
@hadley I have simplified the constructors' code and generalized the behavior on empty inputs. Could you please check if you are ok with it?

File: R/durations.r [hunk 2]
```diff
@@ -242,28 +245,25 @@ setMethod("[[<-", signature(x = "Duration"),
...
+    return(new("Duration", numeric()))
+
   if (!is.numeric(num)) {
     stop(sprintf("First argument to `duration` constructor must be character or numeric. Supplied object of class '%s'", class(num)))
   }
 
   ## qucik check for common wrongdoings: https://github.com/hadley/lubridate/issues/462
   if (class(num)[[1]] %in% c("Interval", "Period"))
-    stop("Interval or Period objects cannot be used as input to 'period' constructor. Plese use 'as.duration'.")
+    stop("Interval or Period objects cannot be used as input to 'duration' constructor. Use 'as.duration' instead.")
```
Comment (hadley) [inline R/durations.r]: If you're touching this, I think you should update to:

```suggestion
    stop("Interval or Period objects cannot be used as input to `duration()` constructor. Use `as.duration()` instead.", call. = FALSE)
```

File: R/periods.r [hunk 3]
```diff
@@ -418,13 +421,15 @@ parse_period <- function(x) {
...
+    return(new("Period", numeric()))
 
   if (!is.numeric(num)) {
     stop(sprintf("First argument to `period` constructor must be character or numeric. Supplied object of class '%s'", class(num)))
   }
 
-  if (is.interval(num))
-    stop("Interval objects cannot be used as input to 'period' constructor. Plese use 'as.period'.")
+  if (class(num)[[1]] %in% c("Interval", "Duration"))
+    stop("Interval or Durations objects cannot be used as input to 'period' constructor. Plese use 'as.period'.")
```
Comment (hadley) [inline R/periods.r]: Should be consistent with error message above.

File: R/periods.r [hunk 4]
```diff
@@ -418,13 +421,15 @@ parse_period <- function(x) {
...
+  if (length(num) == 0)
+    return(new("Period", numeric()))
 
   if (!is.numeric(num)) {
     stop(sprintf("First argument to `period` constructor must be character or numeric. Supplied object of class '%s'", class(num)))
   }
 
-  if (is.interval(num))
-    stop("Interval objects cannot be used as input to 'period' constructor. Plese use 'as.period'.")
+  if (class(num)[[1]] %in% c("Interval", "Duration"))
```
Comment (hadley) [inline R/periods.r]: This will fail for subclasses of Interval and Duration, which is probably not what you want.

Comment (vspinu) [reply R/periods.r]: yes, this is what we use elsewhere for efficiency reasons (I guess). Let me change all of that.

**Comment by vspinu** [comment]
>  I think a merge is a bit better when working on a PR branch because it makes it a bit easier for the other contributor (particularly newer github users) to pull back the changes.

Yeh, it's a trade-off between clean history and ease of use. Good that you are not the "newer github user" ;)

**Comment by hadley** [comment]
We now generally do a "squash and merge" when merging, since maintaining the history within a PR isn't usually that important.

**Pull request merged**